File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/WordConstructorFragment.java
package com.vladimir_x.easyenglishlearn.exercises;

import androidx.lifecycle.ViewModelProvider;
import androidx.databinding.DataBindingUtil;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout.LayoutParams;
import android.widget.Toast;

import com.vladimir_x.easyenglishlearn.R;
import com.vladimir_x.easyenglishlearn.databinding.FragmentWordConstructorBinding;
import com.vladimir_x.easyenglishlearn.model.Word;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static com.vladimir_x.easyenglishlearn.Constants.SELECTED_WORDS;
import static com.vladimir_x.easyenglishlearn.Constants.TRANSLATION_DIRECTION;

public class WordConstructorFragment extends Fragment {

    private FragmentWordConstructorBinding mBinding;
    private WordConstructorViewModel mViewModel;


    @NonNull
    public static Fragment newInstance(ArrayList<Word> selectedWordList,
                                       boolean translationDirection) {
        Bundle args = new Bundle();
        args.putParcelableArrayList(SELECTED_WORDS, selectedWordList);
        args.putBoolean(TRANSLATION_DIRECTION, translationDirection);
        Fragment fragment = new WordConstructorFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private OnClickListener newButtonListener = v -> {
        String letter = ((Button) v).getText().toString();
        mViewModel.onNewButtonClick(letter);
        mBinding.wcfGridContainer.removeView(v);
    };

    @Override
    @SuppressWarnings("ConstantConditions")
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater,
                R.layout.fragment_word_constructor,
                container,
                false);
        mViewModel = new ViewModelProvider(this).get(WordConstructorViewModel.class);
        mBinding.setViewModel(mViewModel);
        subscribeToLiveData(mViewModel);

        if (savedInstanceState == null) {
            boolean translationDirection = getArguments().getBoolean(TRANSLATION_DIRECTION);
            List<Word> wordList = getArguments().getParcelableArrayList(SELECTED_WORDS);
            mViewModel.startExercise(wordList, translationDirection);
        }
        return mBinding.getRoot();
    }

    private void subscribeToLiveData(@NonNull WordConstructorViewModel viewModel) {
        viewModel.getExerciseCloseLiveData().observe(getViewLifecycleOwner(), aVoid -> closeFragment());
        viewModel.getMessageLiveData().observe(getViewLifecycleOwner(), this::showMessage);
        viewModel.getCharArrayLiveData().observe(getViewLifecycleOwner(), this::createButtons);
    }

    private void closeFragment() {
        requireActivity().onBackPressed();
    }

    private void showMessage(int errorsCount) {
        String message;
        if (errorsCount < 0) {
            message = getString(R.string.wrong_answer);
        } else {
            message = getString(R.string.errors_count, errorsCount);
        }
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }

    private void createButtons(List<Character> letters) {
        mBinding.wcfGridContainer.removeAllViews();
        for (char letter : letters) {
            Button button = new Button(getActivity());
            button.setText(String.valueOf(letter));
            button.setOnClickListener(newButtonListener);
        /*button.setTextSize(Objects.requireNonNull(getActivity())
                .getBaseContext()
                .getResources()
                .getDimension(R.dimen.text_size));*/
            mBinding.wcfGridContainer.addView(button, new LayoutParams(150, 150));
        }
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/WordConstructorFragment.kt
package com.vladimir_x.easyenglishlearn.exercises

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.vladimir_x.easyenglishlearn.Constants
import com.vladimir_x.easyenglishlearn.R
import com.vladimir_x.easyenglishlearn.databinding.FragmentWordConstructorBinding
import com.vladimir_x.easyenglishlearn.model.Word
import java.util.ArrayList

class WordConstructorFragment : Fragment() {
    private var viewModel: WordConstructorViewModel? = null
    private var _binding: FragmentWordConstructorBinding? = null
    private val binding get() = _binding!!
    private val newButtonListener = View.OnClickListener { v: View ->
        val letter = (v as Button).text.toString()
        viewModel?.onNewButtonClick(letter)
        binding.wcfGridContainer.removeView(v)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentWordConstructorBinding.inflate(
            inflater,
            container,
            false
        )
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel = ViewModelProvider(this).get(WordConstructorViewModel::class.java)
        subscribeToLiveData()
        if (savedInstanceState == null) {
            val translationDirection =
                arguments?.getBoolean(Constants.TRANSLATION_DIRECTION) ?: true
            val wordList =
                arguments?.getParcelableArrayList<Word>(Constants.SELECTED_WORDS) as List<Word>
            viewModel?.startExercise(wordList, translationDirection)
        }
        binding.wcfBtnClean.setOnClickListener {
            viewModel?.onButtonUndoClick()
        }

        binding.wcfTvQuestion.text = viewModel?.question
        binding.wcfTvAnswer.text = viewModel?.answer
    }

    private fun subscribeToLiveData() {
        viewModel?.exerciseCloseLiveData?.observe(viewLifecycleOwner) { closeFragment() }
        viewModel?.messageLiveData?.observe(viewLifecycleOwner) {
            it?.let { errorsCount: Int ->
                showMessage(errorsCount)
            }
        }
        viewModel?.charArrayLiveData?.observe(
            viewLifecycleOwner
        )
        { letters: List<Char> -> createButtons(letters) }
    }

    private fun closeFragment() {
        requireActivity().onBackPressed()
    }

    private fun showMessage(errorsCount: Int) {
        val message = if (errorsCount < 0) {
            getString(R.string.wrong_answer)
        } else {
            getString(R.string.errors_count, errorsCount)
        }
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    private fun createButtons(letters: List<Char>) {
        binding.wcfGridContainer.removeAllViews()
        for (letter in letters) {
            val button = Button(activity)
            button.text = letter.toString()
            button.setOnClickListener(newButtonListener)
            /*button.setTextSize(Objects.requireNonNull(getActivity())
                .getBaseContext()
                .getResources()
                .getDimension(R.dimen.text_size));*/_binding!!.wcfGridContainer.addView(
                button,
                LinearLayout.LayoutParams(150, 150)
            )
        }
    }

    companion object {
        fun newInstance(
            selectedWordList: ArrayList<Word>,
            translationDirection: Boolean
        ) = WordConstructorFragment().apply {
            arguments = createBundle(selectedWordList, translationDirection)
        }

        private fun createBundle(
            selectedWordList: ArrayList<Word>?,
            translationDirection: Boolean
        ) = Bundle().apply {
            putParcelableArrayList(Constants.SELECTED_WORDS, selectedWordList)
            putBoolean(Constants.TRANSLATION_DIRECTION, translationDirection)
        }
    }
}